File: src/org/redrune/engine/cycle/GameCycleWorker.java
package org.redrune.engine.cycle;

import lombok.Getter;
import org.redrune.engine.SystemManager;
import org.redrune.game.global.World;
import org.redrune.utility.functions.Misc;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

public final class GameCycleWorker implements Runnable {
	
	/**
	 * The timestamp of the end of the last cycle
	 */
	public static long lastCycleTime;
	
	/**
	 * The amount of ticks that have passed since the game started
	 */
	@Getter
	private static int ticksPassed = 0;
	
	/**
	 * The instance of the update sequence
	 */
	private final UpdateSequence updateSequence = new UpdateSequence();
	
	/**
	 * The executor.
	 */
	private final Executor EXECUTOR = Executors.newSingleThreadExecutor();
	
	/**
	 * If the major update worker has started.
	 */
	private boolean started;
	
	public GameCycleWorker() {
	
	}
	
	@Override
	public final void run() {
		while (!SystemManager.shutdown) {
			long currentTime = Misc.currentTimeMillis();
			try {
				updateSequence.fire(World.getPlayers(), World.getNPCs());
			} catch (Throwable e) {
				e.printStackTrace();
			}
			sleepThread(currentTime);
		}
	}
	
	/**
	 * Handles the sleeping of the thread
	 */
	private void sleepThread(long currentTime) {
		lastCycleTime = Misc.currentTimeMillis();
		long sleepTime = 600 + currentTime - lastCycleTime;
		if (sleepTime <= 0) {
			return;
		}
		ticksPassed++;
		try {
			Thread.sleep(sleepTime);
		} catch (InterruptedException e) {
			e.printStackTrace();
		}
	}
	
	/**
	 * Starts the worker
	 */
	public void start() {
		if (started) {
			return;
		}
		started = true;
		EXECUTOR.execute(GameCycleWorker.this);
		System.out.println("Main game cycle worker started.");
	}
}

File: src/main/java/org/redrune/engine/cycle/GameCycleWorker.kt
package org.redrune.engine.cycle

import com.github.michaelbull.logging.InlineLogger
import java.util.concurrent.Executors

class GameCycleWorker {

    /**
     * The instance of the update sequence
     */
    private val updateSequence = UpdateSequence()

    /**
     * The executor.
     */
    private val executor = Executors.newSingleThreadExecutor()

    /**
     * If the major update worker has started.
     */
    private var started = false

    /**
     * Starts the worker
     */
    fun start() {
        if (started) {
            return
        }
        started = true
        executor.execute(updateSequence)
        logger.info { "Main game cycle worker started." }
    }

    companion object {

        private val logger = InlineLogger()

        /**
         * The timestamp of the end of the last cycle
         */
        var lastCycleTime: Long = 0

        /**
         * The amount of ticks that have passed since the game started
         */
        @JvmStatic
        var ticksPassed = 0
    }
}